src/sasa.cpp

```diff
@@ -233,4 +239,208 @@ TEST_CASE("[Faunus] SASAPBC") {
+    };
+
+    for (const auto& cell_offset : cell_offsets) {
+
+        const auto& neighbour_particle_indices = neighour_particles_at(cell_offset);
+
+        for (const auto neighbour_particle_index : neighbour_particle_indices) {
+
+            const auto& particle_j = spc.particles.at(neighbour_particle_index);
+            const auto sasa_radius_j = radii[neighbour_particle_index] + probe_radius;
+            const auto sq_cutoff = (sasa_radius_i + sasa_radius_j) * (sasa_radius_i + sasa_radius_j);
+
+            if (target_index != neighbour_particle_index &&
+                spc.geometry.sqdist(particle_i.pos, particle_j.pos) <= sq_cutoff) {
+
+                const auto dr = spc.geometry.vdist(particle_i.pos, particle_j.pos);
+                neighbours.points.push_back(dr);
+                neighbours.indices.push_back(neighbour_particle_index);
+            }
+        }
+    }
+
+    return neighbours;
+}
+
+template <typename CellList_T>
+std::vector<SASABase::Neighbours>
+SASACellList<CellList_T>::calcNeighbourData(Space& spc, const std::vector<AtomIndex>& target_indices) {
+
+    const auto number_of_indices = target_indices.size();
+    std::vector<SASA::Neighbours> neighbours(number_of_indices);
+
+    for (size_t i = 0; i != number_of_indices; ++i) {
+        neighbours.at(i) = calcNeighbourDataOfParticle(spc, target_indices.at(i));
+    }
+
+    return neighbours;
+}
+template <typename CellList_T> void SASACellList<CellList_T>::update(Space& spc, const Change& change) {
+
+    if (change.everything || change.volume_change) {
+        createCellList(spc);
+    } else if (change.matter_change) {
+        updateMatterChange(spc, change);
+    } else {
+        for (const auto& group_change : change.groups) {
+            const auto& group = spc.groups.at(group_change.group_index);
+            const auto offset = spc.getFirstParticleIndex(group);
+            if (group_change.relative_atom_indices.empty()) {
+                const auto changed_atom_indices = ranges::cpp20::views::iota(offset, offset + group.size());
+                for (const auto changed_index : changed_atom_indices) {
+                    cell_list->updateMemberAt(changed_index,
+                                              spc.particles.at(changed_index).pos + spc.geometry.getLength() / 2.);
+                }
+            } else {
+                const auto changed_atom_indices =
+                    group_change.relative_atom_indices |
+                    ranges::cpp20::views::transform([offset](auto i) { return i + offset; });
+                for (const auto changed_index : changed_atom_indices) {
+                    cell_list->updateMemberAt(changed_index,
+                                              spc.particles.at(changed_index).pos + spc.geometry.getLength() / 2.);
+                }
+            }
+        }
+    }
+}
+
+template <typename CellList_T> void SASACellList<CellList_T>::createCellList(Space& spc) {
```

Comment:
slightly more general and less dependent if this function takes a pair of templated iterators (begin, end) and the geometry.